pkg/template/processor.go

```diff
@@ -0,0 +1,111 @@
+package template
+
+import (
+	"fmt"
+	"math/rand"
+	"time"
+
+	"github.com/go-logr/logr"
+	templatev1 "github.com/openshift/api/template/v1"
+	"github.com/openshift/library-go/pkg/template/generator"
+	"github.com/openshift/library-go/pkg/template/templateprocessing"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/api/errors"
+	errs "github.com/pkg/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"context"
+)
+
+type Processor struct {
+	cl     client.Client
+	scheme *runtime.Scheme
+	log    logr.Logger
+}
+
+func NewProcessor(cl client.Client, scheme *runtime.Scheme, log logr.Logger) *Processor {
+	return &Processor{cl: cl, scheme: scheme, log: log}
+}
+
+func (p Processor) ProcessAndApply(tmplContent []byte, values map[string]string) error {
+	objs, err := p.Process(tmplContent, values)
+	if err != nil {
+		return err
+	}
+
+	return apply(p.cl, objs, p.log)
+}
+
+func apply(cl client.Client, objs []runtime.RawExtension, rq logr.Logger) error {
+	for _, obj := range objs {
+		if obj.Object == nil {
+			rq.Info("template object is nil")
+			continue
+		}
+		gvk := obj.Object.GetObjectKind().GroupVersionKind()
+		rq.Info("processing object", "version", gvk.Version, "kind", gvk.Kind)
+		if err := cl.Create(context.TODO(), obj.Object); err != nil {
+			if errors.IsAlreadyExists(err) {
+				// If client failed to create all resources(few created, few remaining) in the first run, then to avoid
+				// resource already existing errors for later runs.(it's like oc apply, don't fail to exists)
+				continue
```

Comment:
Right.
But `oc apply` actually doesn't recreate objects afaik. It creates new ones and patches the existing ones. So, maybe we can update the existing objects instead of removing/re-creating them?

Reply:
> I am not sure how do we get conflicts,

in my case conflict == response code 409. You will get it when you send a POST request to OS for an object that already exists (which is exactly what the client does)

Yeah, updating of the object will work too, just be sure that the spec is replaced by the new one - to ensure that all previous configuration is removed and replaced/updated by the new one